### DISCUSSION TITLE
How to intercept route changes?

### DISCUSSION THREAD

**Discussion by janhesters** [vercel/next.js#12348] [Help]
Based on same React state (streaming app, so e.g. the user currently streaming), we need to intercept the navigation of the user. The user needs to be prompted, whether he wants to continue leaving the page, or not. If he is sure he wants to leave the page, we need to continue with the navigation. If not, we need to cancel the navigation. How do you do this with Next.js?

I have seen a big discussion about this in this PR https://github.com/zeit/next.js/pull/5377 that (IMO) has been closed prematurely. Is there an official way of achieving the aforementioned behavior? If not, can we make this a feature request?

One suggested solution has been to use custom link tags, but some of our navigation happens inside of Redux Sagas programatically. So we need to somehow be able to act based on events from the Next.js `Router` and cancel if necessary.

## API Example

```js
Router.events.on('beforeRouteChange', () => isStreaming && confirm('Do you really want to leave?'));
```

**Comment by mattboon** [comment]
I recently used [the workaround in this thread](https://github.com/zeit/next.js/issues/2476), but would love to see a proper feature implemented.

```js
Router.events.on('routeChangeStart', () => {
  // Cancel routeChange event by erroring
  // See https://github.com/zeit/next.js/issues/2476
  Router.events.emit('routeChangeError')
  throw `routeChange aborted. This error can be safely ignored - https://github.com/zeit/next.js/issues/2476.`
})
```

**Reply by flippidippi** [reply]
I found that this almost works but behaves weird when using the back and forward browser buttons. Sometimes it'll still change the url but not the page itself.

**Reply by ItsJohnTheDeveloper** [reply]
@flippidippi Same issue is happening to me intermittently. Sometimes I get default browser confirmation window and other times I will get my custom confirmation message

**Reply by fitra36** [reply]
@flippidippi @ItsJohnTheDeveloper is there any solution ?

**Reply by tomasrodathg** [reply]
Late to the party but if you’re using Next’s link don’t intercept the router intercept the window’s popstate event as it’s all Aja and will detect any changes in history which is the base of navigation. Listen to pop state in a layout element and adapt to router changes

https://developer.mozilla.org/en-US/docs/Web/API/Window/popstate_event

**Reply by tomasrodathg** [reply]
Caveat - if you’re trying to intercept before navigation happens you’re better off using intersection observers using an animation library like Motion. Popstate covers post navigation after the Ui updates

**Comment by dvakatsiienko** [comment]
+1 on this topic. It is highly desirable to have route change abort if `false` is returned from an event handle for instance.

```javascript
Router.events.on('beforeHistoryChange', (route: string) => {
    if (route.includes('PRIVATE_ROUTE_NAME')) {
        return false;
    }

    return true;
});

```

**Reply by ghost** [reply]
When false is returned, the url of browser should not be changed.

**Reply by alipetarian** [reply]
I tried that but its not working for me.

`const handleBeforeHistoryChange= (route) => {
    console.log('BEFORE HISTORY CHANGE', route)
    return false
  };

  useEffect(() => {
    console.log("SETUP", router.query, router.asPath);

    // on initial page load.
    handleRouteChange(router?.asPath || "");

    // router.events.on("routeChangeStart", handleRouteChange);
    router.events.on("routeChangeComplete", handleRouteChange);
    router.events.on('beforeHistoryChange',handleBeforeHistoryChange)

    return () => {
      // router.events.off("routeChangeStart", handleRouteChange)
      router.events.off("routeChangeComplete", handleRouteChange)
      router.events.off('beforeHistoryChange',handleBeforeHistoryChange)
    };
  }, []);
`

**Reply by ghost** [reply]
Sorry, it is the return value of beforePopState method and it worked for me.
When return false, it is need to rollback the browser history.

```js
   const lastHistoryStateRef = useRef(global.history?.state);

    router.beforePopState(({ url }) => {
      if (!canPopup(url, 'POP')) {
        const lastHistoryState = lastHistoryStateRef.current;
        if (lastHistoryState && window.history.state && lastHistoryState.idx !== window.history.state.idx) {
          window.history.go(lastHistoryState.idx < window.history.state.idx ? -1 : 1);
        }
        return false;
      }
      return true;
    });
```

**Comment by jflayhart** [comment]
Is there anything in NextJS that reliably handles this type of behavior? I'm trying to do something similar to what the `history` library allows:

```
history.block(location => {
  if (location.pathname === '/') return false
  else return true
})
```

**Comment by janhesters** [comment]
The solution we ended up using was globally listening to events. We created a `RouteInterceptor` component, that was added to our `Layout` component. We spy on the events using `composedPath`.

```js
const handleClick = event => {
  if ([/* ... blocked paths ... */].includes(router.pathname)) {
    const eventWillRedirect = event
      .composedPath()
      .filter(el => el.nodeName === 'A');

    // interceptionConfig defines custom rules and settings for route interception
    const config = interceptionConfig();

    if (eventWillRedirect.length && config.shouldIntercept) {
      const path = eventWillRedirect[0].attributes.href.value;
      // custom logic for our purposes
      const areaPrefix = `/areas/${areaSlug}`;
      const urlIsInsideArea = path.indexOf(areaPrefix) === 0;

      if (urlIsInsideArea) {
        event.preventDefault();
        // opens route interceptor modal which has a button to continue or to abort
        openModal({ contentName: 'routeInterceptor' });
        setInterceptedRoute(path);
        setInterceptedAsPath(getGenericPath(path));
        setInterceptionReason(config.reason);
      }
    }
  }
};
```

And then set up the listener in a `useEffect`:

```js
useEffect(() => {
  document.addEventListener('click', handleClick, { capture: true });

  return () =>
    document.removeEventListener('click', handleClick, { capture: true });
}, [/* ... deps from Redux */, handleClick]);
```

**Comment by Fensterbank** [comment]
Thanks for the discussion and the possible workarounds.
Using your examples I built a custom hook, which can be used with `useLeavePrevention()` in any component.

```javascript
const handleWindowClose = (e) => {
  e.preventDefault();
  return (e.returnValue = 'Are you sure you want to leave?');
};

let leaveConfirmed = false;

export const useLeavePrevention = () => {
  const router = useRouter();

  // Use beforeunload to prevent closing the tab, refreshing the page or moving outside the Next app
  useEffect(() => {
    window.addEventListener('beforeunload', handleWindowClose);

    return () => {
      window.removeEventListener('beforeunload', handleWindowClose);
    };
  });

  // Use routeChangeStart to prevent navigation inside of the Next app
  // Uses a module variable to bypass the confirm, otherwise we would be in a loop
  router.events.on('routeChangeStart', () => {
    if (leaveConfirmed) return;

    if (window.confirm('Are you sure you want to leave?'))
      leaveConfirmed = true;
    else {
      router.events.emit('routeChangeError');
      throw 'routeChange aborted.';
    }
  });

  // Set the module variable to false on component mount
  useEffect(() => {
    leaveConfirmed = false;
  }, []);
};
```
I don't have any parameters so it always asks and this global module variable could bring up some issues, but for my case it works and you get the rough idea.

**Reply by christianwish** [reply]
@Fensterbank 
Inspired by your last comment I could create a custom hook that shows a custom Prompt:

```ts
import { useEffect, useReducer, useState } from 'react';
import { useRouter } from 'next/router';

export const useDataReducer = (reducer, initialState = {}) => {
    const router = useRouter();
    const [changed, setChanged] = useState(false);
    const [{nextRoute, confirmed }, setNextRoute] = useState({ nextRoute: null, confirmed: false });
    const [state, dispatch] = useReducer(reducer, initialState);

    const toNextRoute = () => setNextRoute({ nextRoute, confirmed: true });
    const stayHere = () => setNextRoute({ nextRoute: null, confirmed: false });

    useEffect(() => {
        const onRouteChangeStart = (route) => {
            if (!(changed)) return undefined;

            setNextRoute({ nextRoute: route, confirmed: false });
            router.events.emit('routeChangeError');
            throw 'routeChange aborted.';
        };

        const cleanUpFunction = () => router.events.off('routeChangeStart', onRouteChangeStart);

        if (nextRoute && confirmed) {
            router.push(nextRoute)
            return cleanUpFunction;
        }

        router.events.on('routeChangeStart', onRouteChangeStart);

        return cleanUpFunction
    }, [nextRoute, confirmed, state]);

    const Prompt = () => {
        if (!(nextRoute)) return null;
        return (
            <>
                <p>Are you sure you want to leave?</p>
                <button onClick={toNextRoute}>yes</button>
                <button onClick={stayHere}>no</button>
            </>
        );
    }

    const dispatchWrap = (action) => {
        if (!changed) setChanged(true);
        dispatch(action);
    }

    const clear = () => setChanged(false);

    return [state, dispatchWrap, Prompt, clear];
  };

```
Usage:

```ts
const [state, dispatch, Prompt, clear] = useDataReducer(reducer, initialState);

return <>
   <Prompt />
  <MyForm>
</>
```
When you use dispatch the first time and then try to change the route , Prompt will render "yes" and "no" buttons.
Thanks for your inspiration.

**Reply by DanWttvn** [reply]
I leave another version based on @christianheyn solution but without the use of reducers and with a prop that conditions if the confirmation is necessary or not

```typescript 
export const useConfirmLeave = ({ shouldWarn }: { shouldWarn: boolean }) => {
  const router = useRouter();
  const [hasConfirmed, setHasConfirmed] = useState(false);
  const [navigationConfig, setNavigationConfig] = useState<{
    nextRoute: string | null;
    isModalOpen: boolean;
  }>({
    nextRoute: null,
    isModalOpen: false,
  });

  // Use beforeunload to prevent closing the tab, refreshing the page or moving outside the Next app
  useEffect(() => {
    const handleWindowClose = (e: BeforeUnloadEvent) => {
      if (!shouldWarn) return;
      e.preventDefault();
      const event = e || window.event;
      return (event.returnValue = 'Are you sure you want to leave?');
    };

    window.addEventListener('beforeunload', handleWindowClose);
    return () => {
      window.removeEventListener('beforeunload', handleWindowClose);
    };
  }, [shouldWarn]);

  // Use routeChangeStart to prevent navigation inside of the Next app
  useEffect(() => {
    const onRouteChangeStart = (route: string) => {
      if (!shouldWarn || hasConfirmed) return;
      else {
        setNavigationConfig({
          nextRoute: route,
          isModalOpen: true,
        });
        router.events.emit('routeChangeError');
        // eslint-disable-next-line @typescript-eslint/no-throw-literal
        throw 'navigation aborted';
      }
    };
    router.events.on('routeChangeStart', onRouteChangeStart);
    const cleanUp = () =>
      router.events.off('routeChangeStart', onRouteChangeStart);

    if (hasConfirmed) {
      if (!navigationConfig.nextRoute) return;
      void router.push(navigationConfig.nextRoute);
      return cleanUp;
    }

    return cleanUp;
  }, [navigationConfig, hasConfirmed, router, shouldWarn]);

  const ConfirmLeaveModal = () => (
    <DataNotSavedModal
      isOpen={navigationConfig.isModalOpen}
      onConfirm={() => setHasConfirmed(true)}
      onCancel={() => {
        setNavigationConfig({
          nextRoute: null,
          isModalOpen: false,
        });
      }}
    />
  );

  return {
    ConfirmLeaveModal,
  };
};
```
Usage
```typescript
const { ConfirmLeaveModal } = useConfirmLeave({
  shouldWarn: formState.isDirty && !formState.isSubmitted,
});

return (
  <>
    <ConfirmLeaveModal />
    <Form />
  </>
);
```

**Reply by YYCoder** [reply]
best solution ever, thanks a lot!!!

**Reply by kshashank231** [reply]
The custom Dialog is not working anymore in Nextjs 13, as the router events are removed from this version.

**Comment by GorkemSahin** [comment]
Might be an alternative: https://github.com/vercel/next.js/issues/2476#issuecomment-843311387

**Comment by eric-burel** [comment]
Hi, I am working on a more robust alternative. I don't like the idea of coupling too much this feature with the framework. Here is what I am implementing currently ~~[link to the code](https://gist.github.com/eric-burel/75301e2ba01967e0ec909ddf7fae5100)~~ [Updated code here](https://github.com/VulcanJS/vulcan-npm/tree/main/packages/react-ui/components/form/useBlockTransition)

Basically, I split the process in multiple steps:
- there is a generic hook that handles standard browser level blocking/unblocking as usual + the confirmation message, this is what I've linked
- it triggers 2 special events, "blocktransition" and "unblocktransition"
- in your Next.js app, you can listen those events to setup the blocking for SPA transition, so using Next own routers. There you can use the code that has been demonstrated or Next specific events.
- But you could also listen to those event in a React Router based app => the "blocktransition" event allows decoupling the browser part , your app, and the actual router implementation provided by the framework.

If you switch framework in 5 years, no problem, there is no coupling (in our case we strive to support any React based framework so Gatsby, Next, our legacy Meteor version of Vulcan).

**Reply by eric-burel** [reply]
Can you please elaborate the thumbs down? I've updated the link to the updated code as I've fixed a few issues in the existing one.

**Reply by adams-family** [reply]
Hi @eric-burel - new to this thread, I was wondering if the route change intercepting has been implemented in the meantime? I think it's crucial to prevent user's data loss on accidental back-button gesture (very easy to do in Chrome). Thanks!

**Comment by prakhar-parashar-au7** [comment]
Heyy, I'm using next.js, this code works fine when I use it in [path].js , and when route is being changed from '/something' to '/somethingelse', 
but it doesn't work when I put it in index.js, and when I'm going from '/' to '/somethingElse'. Is there any reason why it doesn't work in index.js?

**Reply by chloekim02** [reply]
it's been a while, did you find a solution?

**Comment by yeunoia** [comment]
I am trying to go other url when there is no history
but this doesn't work 
any ideas? 
```js
  useEffect(() => {
    //console.log(historyIndex);
    const beforeHistoryChange = () => {
      if (historyIndex === 0) {
        router.push('/campaigns2');
      }
    };
    router.events.on('beforeHistoryChange', beforeHistoryChange);

    return () => {
      router.events.off('beforeHistoryChange', beforeHistoryChange);
    };
  }, []);

```

**Reply by joe-shajan** [reply]
how can i do this in latest app router?

**Comment by prince272** [comment]
Support for App Router

**Reply by ray-silicon-rhino** [reply]
yeah please, I'm having the issue of having an event listener in routeChangeStart, is there a way to do this in the new appRouter?

**Reply by johannbuscail** [reply]
any way to do this now ?

**Comment by ihsanmaulanahamdani** [comment]
How to use router.beforePopState in NextJs 13/14. Please help, Thank you.

**Comment by cgfeel** [comment]
## Intercept route changes at the NextJS app router mode

Demo：[[codeSondbox](https://codesandbox.io/p/devbox/next-block-leaving-yf9nlq?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clq1ovh9a000b3b6lujkvkri7%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clq1ovh9900023b6l4kn66bas%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clq1ovh9900083b6lm7e16teb%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clq1ovh99000a3b6lqm95cmz6%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clq1ovh9900023b6l4kn66bas%2522%253A%257B%2522id%2522%253A%2522clq1ovh9900023b6l4kn66bas%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clq1p4cpn00023b6ld60iy8s9%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A4%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A4%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fapp%252Fform.css%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clq1p4cpn00023b6ld60iy8s9%2522%257D%252C%2522clq1ovh99000a3b6lqm95cmz6%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clq1ovh9900093b6l2s8mq53w%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522path%2522%253A%2522%252Fproxy%252Fstore%2522%257D%255D%252C%2522id%2522%253A%2522clq1ovh99000a3b6lqm95cmz6%2522%252C%2522activeTabId%2522%253A%2522clq1ovh9900093b6l2s8mq53w%2522%257D%252C%2522clq1ovh9900083b6lm7e16teb%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clq1ovh9900033b6lp2vrvqp7%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%252C%257B%2522id%2522%253A%2522clq1ovh9900043b6ll8bpphgd%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522build%2522%257D%252C%257B%2522id%2522%253A%2522clq1ovh9900053b6lxnemchtd%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%257D%252C%257B%2522id%2522%253A%2522clq1ovh9900063b6lqbse2rr4%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522lint%2522%257D%252C%257B%2522id%2522%253A%2522clq1ovh9900073b6lobmjoxn5%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522install%2522%257D%255D%252C%2522id%2522%253A%2522clq1ovh9900083b6lm7e16teb%2522%252C%2522activeTabId%2522%253A%2522clq1ovh9900033b6lp2vrvqp7%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)]

https://github.com/cgfeel/next.v2/assets/578141/fe1e7f24-054a-4f56-892c-5345ac177a75

source code: https://github.com/cgfeel/next.v2/tree/master/routing-file/src/app/leaving/proxy

Use this Provider in your layout at the app root directory：https://github.com/cgfeel/next.v2/blob/master/routing-file/src/components/proxyProvider/index.tsx

```tsx
'use client'

import { usePathname, useSearchParams } from "next/navigation";
import Script from "next/script";
import { FC, PropsWithChildren, createContext, useEffect, useState } from "react";

const ProxyContext = createContext<ProxyInstance>([undefined, () => {}]);

const ProxyProvider: FC<PropsWithChildren<{}>> = ({ children }) => {
    const [tips, setTips] = useState<string|undefined>();
    const msg = tips === undefined ? tips : (tips||'Are you sure want to leave this page?');

    const pathname = usePathname();
    const searchParams = useSearchParams();

    const url = [pathname, searchParams].filter(i => i).join('?');
    useEffect(() => {
        setTips(undefined);
    }, [url, setTips]);

    useEffect(() => {
        const handleBeforeUnload = (event: BeforeUnloadEvent) => {
            if (msg === undefined) return msg;
            
            event.preventDefault();
            event.returnValue = msg;
            
            return msg;
        };

        const script = document.getElementById('proxy-script');
        if (script) {
            script.dataset.msg = msg||'';
            script.dataset.href = location.href;
        }

        window.addEventListener("beforeunload", handleBeforeUnload);
        return () => {
            window.removeEventListener("beforeunload", handleBeforeUnload);
        }
    }, [msg]);

    return (
        <ProxyContext.Provider
            value={[msg, setTips]}
        >
            <Script
                strategy="afterInteractive"
                id="proxy-script"
                dangerouslySetInnerHTML={{
                    __html: `(() => {
                        const originalPushState = history.pushState.bind(history);
                        let currentPoint = 0;
                        let point = 0;
                        window.history.pushState = function(state, title, url) {
                            state.point = ++point;
                            currentPoint = point;
                            originalPushState(state, title, url);
                        };
                        const originalReplaceState = history.replaceState.bind(history);
                        window.history.replaceState = function(state, title, url) {
                            state.point = currentPoint;
                            originalReplaceState(state, title, url);
                        };
                        window.addEventListener('popstate', function (event) {
                            const { state: nextState } = event;
                            const isback = currentPoint > nextState.point;

                            currentPoint = nextState.point;

                            const script = document.getElementById('proxy-script');
                            if (!script || location.href === script.dataset.href) return;

                            const msg = script.dataset.msg||'';
                            const confirm = msg == '' ? true : window.confirm(msg);
                            if (!confirm) {
                                event.stopImmediatePropagation();
                                isback ? history.forward() : history.back();
                            }
                        });
                    })()`,
                }}
            ></Script>
            {children}
        </ProxyContext.Provider>
    );
};

export type ProxyInstance = [
    string|undefined, (tips?: string) => void
]

export { ProxyContext };

export default ProxyProvider;
```

**Reply by mrzarkovic** [reply]
amazing, but it also works!

**Comment by TomRaaff** [comment]
For those finding this answer after 2025:
Next now has an 'official' way to handle this in their documentation.
https://nextjs.org/docs/app/api-reference/components/link#blocking-navigation
it doesn't handle cases like navigating to an entirely different website, but combined with the answers above, you can figure it out.

**Reply by thaoms** [reply]
Next.js solution is pretty poor. Having to wrap all your Links as a custom component, and not taking in account, closing tabs and such.

Why not just this?

```typescript
"use client";

import { useEffect } from "react";

export function useUnsavedChanges(isDirty: boolean) {
  useEffect(() => {
    if (!isDirty) return;

    const beforeUnload = (e: BeforeUnloadEvent) => {
      e.preventDefault();
      e.returnValue = ""; // legacy support
    };

    const onClick = (e: MouseEvent) => {
      const target = e.target as HTMLElement;
      const anchor = target.closest("a");

      if (!anchor) return;
      if (!anchor.href) return;
      if (anchor.target === "_blank") return;
      if (anchor.hasAttribute("download")) return;

      const ok = window.confirm(
        "You have unsaved changes. Are you sure you want to leave?",
      );

      if (!ok) {
        e.preventDefault();
        e.stopImmediatePropagation();
      }
    };

    window.addEventListener("beforeunload", beforeUnload);
    document.addEventListener("click", onClick, true);

    return () => {
      window.removeEventListener("beforeunload", beforeUnload);
      document.removeEventListener("click", onClick, true);
    };
  }, [isDirty]);
}

```

and use for example like this : `useUnsavedChanges(form.formState.isDirty);`
